portal/src/graphql/portal/EditOAuthClientScreen.tsx

```diff
@@ -0,0 +1,190 @@
+import React, {
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+import { useNavigate, useParams } from "react-router-dom";
+import deepEqual from "deep-equal";
+import produce from "immer";
+import { Text, TextField } from "@fluentui/react";
+import { Context, FormattedMessage } from "@oursky/react-messageformat";
+
+import NavBreadcrumb, { BreadcrumbItem } from "../../NavBreadcrumb";
+import ShowError from "../../ShowError";
+import ShowLoading from "../../ShowLoading";
+import ButtonWithLoading from "../../ButtonWithLoading";
+import NavigationBlockerDialog from "../../NavigationBlockerDialog";
+import ModifyOAuthClientForm from "./ModifyOAuthClientForm";
+import { useAppConfigQuery } from "./query/appConfigQuery";
+import { useUpdateAppConfigMutation } from "./mutations/updateAppConfigMutation";
+import { OAuthClientConfig, PortalAPIAppConfig } from "../../types";
+import { clearEmptyObject } from "../../util/misc";
+
+import styles from "./EditOAuthClientScreen.module.scss";
+
+interface EditOAuthClientFormProps {
+  clientConfig: OAuthClientConfig;
+  rawAppConfig: PortalAPIAppConfig;
+}
+
+const EditOAuthClientForm: React.FC<EditOAuthClientFormProps> = function EditOAuthClientForm(
+  props: EditOAuthClientFormProps
+) {
+  const { clientConfig: clientConfigProps, rawAppConfig } = props;
+  const { renderToString } = useContext(Context);
+  const { appID } = useParams();
+  const navigate = useNavigate();
+
+  const {
+    updateAppConfig,
+    loading: updatingAppConfig,
+    error: updateAppConfigError,
+  } = useUpdateAppConfigMutation(appID);
+
+  const initialClientConfig = useMemo(() => {
+    return {
+      ...clientConfigProps,
+      post_logout_redirect_uris:
+        (clientConfigProps.post_logout_redirect_uris ?? []).length > 0
+          ? clientConfigProps.post_logout_redirect_uris
+          : undefined,
+    };
+  }, [clientConfigProps]);
+
+  const [clientConfig, setClientConfig] = useState<OAuthClientConfig>(
+    initialClientConfig
+  );
+  const [submittedForm, setSubmittedForm] = useState(false);
+
+  const onClientConfigChange = useCallback(
+    (newClientConfig: OAuthClientConfig) => {
+      setClientConfig(newClientConfig);
+    },
+    []
+  );
+
+  const onSaveClick = useCallback(() => {
+    const newAppConfig = produce(rawAppConfig, (draftConfig) => {
+      const clients = draftConfig.oauth!.clients!;
+      const clientConfigIndex = clients.findIndex(
+        (client) => client.client_id === clientConfig.client_id
+      );
+      clients[clientConfigIndex] = clientConfig;
+
+      clearEmptyObject(draftConfig);
+    });
+
+    updateAppConfig(newAppConfig)
+      .then((result) => {
+        if (result != null) {
+          setSubmittedForm(true);
+        }
+      })
+      .catch(() => {});
+  }, [clientConfig, updateAppConfig, rawAppConfig]);
+
+  useEffect(() => {
+    if (submittedForm) {
```

Comment:
It seems that the only use of `submittedForm` is here. Why don't we just call `navigate` in line 82? Using an useEffect in this way sounds a bit indirect.

Reply:
This is for making sure the state is set before calling navigate, as `setState` in functional component does not have callback @louischan-oursky

Reply:
A second thought. I guess we should stay on the page even after saving.